### PR TITLE
Close #70. Add localStorage for state retention in Timer and Stopwatch

### LINKED ISSUE #70: Add localStorage for state retention 🎯1/2 day
The app should retain its current state between page reloads.

This includes things like:

* The current time left in the Timer (whether its running or not)
* The current time of the Stopwatch (whether its running or not)

### PR DESCRIPTION
**Pull request by kaxcode**
- Add logic for state to be set with localStorage when there is a refresh/update.
- Add logic for localStorage to be set to 0 when unmounted.
- Add jest-localstorage-mock package for testing
- Remove afterEach function in TimerCotainer.test.js file because it was affecting the localStorage mock.
- Add jest.resetAllMocks() to 3 individual tests that need the afterEach function to run mocks reset.

### CODE REVIEW

File: src/components/Stopwatch/__tests__/StopwatchContainer.test.js [hunk 1]
```diff
@@ -96,4 +97,17 @@ describe('StopwatchContainer', () => {
...
+  describe('when unmounted', () => {
+    it('sets localStorage to 0', () => {
+      // Arrange
+      wrapper.setState({ counter: 0 });
+
+      // Act
+      wrapper.instance().componentWillUnmount();
+
+      // Assert
+      expect(wrapper.state().counter).toEqual(0);
```
Comment (dlindahl) [inline src/components/Stopwatch/__tests__/StopwatchContainer.test.js]: Isn't this checking the _component's_ state, not `localStorage`'s?

File: src/components/Stopwatch/__tests__/StopwatchContainer.test.js [hunk 2]
```diff
@@ -96,4 +97,17 @@ describe('StopwatchContainer', () => {
       expect(wrapper.state().counter).not.toBe(0);
     });
   });
+
```
Comment (dlindahl) [inline src/components/Stopwatch/__tests__/StopwatchContainer.test.js]: Shouldn't there be some tests to assert the state of `localStorage` when updating or that state is retrieved from `localStorage` when mounting?

File: src/components/Stopwatch/__tests__/StopwatchContainer.test.js (hunk 1 above)
Comment (kaxcode) [reply src/components/Stopwatch/__tests__/StopwatchContainer.test.js]: The component's state is set by the localStorage. So it's the same. I couldn't test setting and getting from localStorage because it's a mock and they're all an empty function.

Comment (kaxcode) [reply src/components/Stopwatch/__tests__/StopwatchContainer.test.js]: I updated this to check localStorage with a `.toHaveBeenLastCalledWith` assertion.

File: src/components/Stopwatch/__tests__/StopwatchContainer.test.js (hunk 2 above)
Comment (kaxcode) [reply src/components/Stopwatch/__tests__/StopwatchContainer.test.js]: I added a test to cover `componentDidUpdate()`.

File: package.json [hunk 3]
```diff
@@ -24,7 +25,7 @@
     "start": "react-scripts start",
     "build": "react-scripts build",
     "test:lint:js": "eslint --fix --ignore-path .gitignore --ext=js --ext=jsx .",
-    "test:unit": "react-scripts test  --coverage --env=jsdom",
+    "test:unit": "react-scripts test --coverage --env=jsdom",
```
Comment (dlindahl) [inline package.json]: Watch commits like this. I get that its a small whitespace change, but it doesn't have anything to do with the PR. This should be a standalone commit "Fix whitespace in `test:unit` task" in this branch or in `master`

Comment (kaxcode) [reply package.json]: Do you want me to change it back?

File: src/components/Stopwatch/__tests__/StopwatchContainer.test.js (hunk 1 above)
Comment (dlindahl) [reply src/components/Stopwatch/__tests__/StopwatchContainer.test.js]: Sure you can test setting and getting from `localStorage`. From the `jest-localstorage-mock` docs:

> Check that localStorage calls were not made when they shouldn't have been.
> 
```
test('should not have saved to localStorage', () => {
  const KEY = 'foo',
    VALUE = 'bar';
  dispatch(action.notIdempotent(KEY, VALUE));
  expect(localStorage.setItem).not.toHaveBeenLastCalledWith(KEY, VALUE);
  expect(Object.keys(localStorage.__STORE__).length).toBe(0);
});
```

Some good test scenarios here would be:

```
describe('StopwatchContainer', () => {
  describe('#componentDidMount', () => {
    it('sets an initial counter value from localstorage when available', () => {
      // set localstorage
      // mount component
      // assert state matches value from LS
    });
    it('sets a sane default counter value', () => {
      // clear LS or set `counter` to null
      // mount component
      // assert state matches the default value
    });
  });
  describe('#componentDidUpdate', () => {
    it('updates localstorage with the counter state', () => {
      // mount component
      // update component
      // assert LS matches component state
    });
  });
  describe('#componentWillUnmount', () => {
    it('resets the stored state in localstorage', () => {
      // set LS
      // mount component
      // unmount component
      // assert LS is reset
    });
  });
});
```

File: src/components/Stopwatch/StopwatchContainer.js [hunk 4]
```diff
@@ -9,6 +9,24 @@ export default class StopwatchContainer extends React.Component {
 
   timer = null;
 
+  componentDidMount = () => {
+    // Looks for localStorage to use
+    const localStorageRef = localStorage.getItem('counter');
+    if (localStorageRef) {
```
[the remark this answers is not in the document]
Comment (dlindahl) [reply src/components/Stopwatch/StopwatchContainer.js]: I meant a default value if `getItem` does not return anything. Something like this: `localStorage.getItem('counter') || 0`.

That way the `if` is unnecessary.

File: src/components/Stopwatch/__tests__/StopwatchContainer.test.js (hunk 1 above)
Comment (kaxcode) [reply src/components/Stopwatch/__tests__/StopwatchContainer.test.js]: @dlindahl I can check setItem with `.toHaveBeenLastCalledWith` like I currently have in code. I can't check .getItem and compare to the state.

Comment (kaxcode) [reply src/components/Stopwatch/__tests__/StopwatchContainer.test.js]: @dlindahl I don't think I'm explaining this well. I can only use `jest-localstorage-mock` to see that calls to localStorage where made, but I can't configure or setup .`setItem` and `.getItem`.

**Comment by kaxcode** [comment]
@dlindahl regarding my latest changes excluding what you suggested. 

it('updates localstorage with the counter state’)- I would but I can’t set LS. I can only see that it was called and with what value & key. 

it('sets a sane default counter value’)- Under componentDidMount the state will be set to 0 or the LS value saved (if updated). 0 is already the default state. We weren’t testing that state was 0 before the LS implementation.

it('sets an initial counter value from localstorage when available’)- Under componentDidMount LS cannot be set because every refresh would reset the timer and LS to “sane default". When you refresh a page you mount and update. This isn’t how the feature works.

File: src/components/Stopwatch/__tests__/StopwatchContainer.test.js (hunk 1 above)
Comment (dlindahl) [reply src/components/Stopwatch/__tests__/StopwatchContainer.test.js]: > The `__STORE__` attribute of `localStorage.__STORE__` or `sessionStorage.__STORE__` is made available for you to directly access the storage object if needed.

You don't need to mock `setItem` or `getItem`. The docs suggest that you can check the state of the mocked store with `localStorage.__STORE__`

```
localStorage.setItem('counter, '25');
wrapper.instance().componentWillUnmount();
expect(localStorage.__STORE__.counter).toEqual(0);
```

Does that work?

As a test author, I don't really care that `setItem` is called, I only care about what the store contains. How that data gets in there is an implementation detail that is not important in the context of this test. If you didn't have a way to inspect the store's state, then mocking would make sense.

Comment (kaxcode) [reply src/components/Stopwatch/__tests__/StopwatchContainer.test.js]: image
image

**Comment by dlindahl** [comment]
> it('updates localstorage with the counter state’)- I would but I can’t set LS. I can only see that it was called and with what value & key.

I wrote a test for this in `8b86a5d`

> it('sets a sane default counter value’)- Under componentDidMount the state will be set to 0 or the LS value saved (if updated). 0 is already the default state. We weren’t testing that state was 0 before the LS implementation.

The point of the test is to ensure that if `localStorage` is not available or is corrupted (non-number stored in `counter`), then the initial value is set to 0 (instead of `NaN`). There was no test for this before because there was no logic involved when initializing state. Now there is so it must be tested. I wrote a test for this in `8b86a5d`. 

> it('sets an initial counter value from localstorage when available’)- Under componentDidMount LS cannot be set because every refresh would reset the timer and LS to “sane default". When you refresh a page you mount and update. This isn’t how the feature works.

"it sets an initial counter value **from** localstorage when available" - I'm not suggesting that `localStorage` itself is updated on mount, but that the current value in `localStorage` is used to initialize the component instance. I wrote a test for this in `8b86a5d`
